File: app/src/main/java/com/extremedesign/typeracer/DataRepository/IDataBaseImplementations/DatabaseDataSource.java
package com.extremedesign.typeracer.DataRepository.IDataBaseImplementations;

import android.content.Context;

import com.extremedesign.typeracer.DataRepository.RoomDatabase.TypeRacerDatabase;
import com.extremedesign.typeracer.listener.JobWorker;
import com.extremedesign.typeracer.listener.ProfileImageListener;
import com.extremedesign.typeracer.model.ProfileImage;
import java.util.List;

public class DatabaseDataSource implements IDataSource {
    private Context context;

    public DatabaseDataSource() {
    }
    public DatabaseDataSource(Context context) {
        this.context=context;
    }


    @Override
    public void insertDataProfileImage(final ProfileImage profileImage, final JobWorker listener) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                TypeRacerDatabase.getDatabase(context).typeRacerDAO().insertProfileImage(profileImage);
                listener.jobFinished(true);
            }
        }).start();
    }

    @Override
    public void insertDataProfileImages(final List<ProfileImage> profileImages) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                TypeRacerDatabase.getDatabase(context).typeRacerDAO().insertProfileImages(profileImages);
            }
        }).start();

    }

    @Override
    public void getDataProfileImages(final ProfileImageListener listener) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                List<ProfileImage> profileImages = TypeRacerDatabase.getDatabase(context).typeRacerDAO().getAllProfileImages();
                listener.getImages(profileImages);
            }
        }).start();

    }

    @Override
    public void getProfileImageByName(final String name, final ProfileImageListener listener) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                List<ProfileImage> profileImages = TypeRacerDatabase.getDatabase(context).typeRacerDAO().getProfileImageByName(name);
                listener.getImages(profileImages);
            }
        }).start();

    }

//    @Override
//    public void getData(String amount, String category, String difficulty, String type, CustomListenerRep listenerRep) {
//        List<Question> questions= QuestionDatabase.getQuestionDatabase(context).noteDao().getAllQuestionsByCategory(category);
//        Collections.shuffle(questions);
//        listenerRep.RepReturn(questions.subList(0,10));
//    }
//    @Override
//    public void insertData(List<Question> questionList) {
//        QuestionDatabase.getQuestionDatabase(context).noteDao().insertNotes(questionList);
//        Category category=new Category(questionList.get(0).getCategory(),questionList.size());
//        QuestionDatabase.getQuestionDatabase(context).noteDao().insertCategory(category);
//
//    }
//
//    @Override
//    public void getListOfCategories(CategoryListener listener) {
//        listener.getAllCategories(QuestionDatabase.getQuestionDatabase(context).noteDao().getAllCategories());
//    }




//
//    @Override
//    public void getData(ListenerRepositoryDataSource listenerRep) {
//     DataBaseHelper db= AppSingleton.getINSTANCE(context).getDbHelper();
//     List<Question>questions=db.getAllQuestions();
//        Log.e("DATABASE_QUESTION_COUNT",String.valueOf(questions.size()));
//     listenerRep.sendDataToRepository(questions);
//        db.close();
//    }
//
//    @Override
//    public void insertData(List<Question> questionList) {
//        DataBaseHelper db= AppSingleton.getINSTANCE(context).getDbHelper();
//            for(Question question:questionList){
//                db.insertQuestion(question);
//            }
//            db.close();
//    }
}

File: app/src/main/java/com/extremedesign/typeracer/data_repository/i_database_implementations/DatabaseDataSource.kt
package com.extremedesign.typeracer.data_repository.i_database_implementations

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.extremedesign.typeracer.data_repository.room_database.TypeRacerDatabase
import com.extremedesign.typeracer.listener.JobWorker
import com.extremedesign.typeracer.listener.ProfileImageListener
import com.extremedesign.typeracer.model.ProfileImage
import com.extremedesign.typeracer.model.User

class DatabaseDataSource(private val context: Context) : IDataSource {
    override fun insertUser(user: User?) {
        Thread { TypeRacerDatabase.getDatabase(context).typeRacerDAO().insertUser(user) }.start()
    }

    override fun getCurrentUser(uid: String?): LiveData<User?> {
        val data = MutableLiveData<User>()
        Thread {
            data.setValue(TypeRacerDatabase.getDatabase(context).typeRacerDAO().getUserByUid(uid))
        }.start()
        return data
    }

    override fun insertDataProfileImage(profileImage: ProfileImage?, listener: JobWorker?) {
        Thread {
            TypeRacerDatabase.getDatabase(context).typeRacerDAO().insertProfileImage(profileImage)
            listener!!.jobFinished(true)
        }.start()
    }

    override fun insertDataProfileImages(profileImages: List<ProfileImage?>?) {
        TODO("Not yet implemented")
    }

    override fun getDataProfileImages(listener: ProfileImageListener?) {
        Thread {
            val profileImages = TypeRacerDatabase.getDatabase(context).typeRacerDAO().getAllProfileImages()
            listener!!.getImages(profileImages)
        }.start()
    }

    override fun getProfileImageByName(name: String?, listener: ProfileImageListener?) {
        Thread {
            val profileImages = TypeRacerDatabase.getDatabase(context).typeRacerDAO().getProfileImageByName(name)
            listener!!.getImages(profileImages)
        }.start()
    }
}